ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+                                                 TimeUnit.MILLISECONDS);
+
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
+        return;
+      }
+      for (long ts : heartbeats.keySet()) {
+
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - configs.heartbeatSendIntervalMs > prev) {
+          missedCount = (ts - prev - 1) / configs.heartbeatSendIntervalMs;
+        } else {
+          missedCount = 0;
+        }
+        prev = ts;
+      }
+      // Check frame from last received heartbeat to window end
+      if (windowEnd - prev - 1 > 0) {
+        missedCount = (windowEnd - prev - 1) / configs.heartbeatSendIntervalMs;
+      }
```

Comment:
I'm not 100% sure I understand this algorithm, but could you just:
```java
int heartbeatsInPeriod = heartbeats.subMap(windowStart, windowEnd).size();
int expectedBeats = (windowEnd - windowStart) / configs.heartBeatSendIntervalMs;
int missed = expectedBeats - heartbeatsInPeriod;
```

Reply:
What you suggest is a different algorithm. We want to count consecutive missed heartbeats and reset the count when we have received heartbeats. It's not enough to just count how many heartbeats we missed in the window as a server may have missed > THRESHOLD but not consecutive which doesn't constitute it as down.

Reply:
got it that makes sense - can you please add an inline comment describing that?